rust/agents/relayer/src/merkle_tree_builder.rs

```diff
@@ -8,20 +8,63 @@ use color_eyre::eyre::Result;
 use ethers::core::types::H256;
 use std::fmt::Display;
 
-use tracing::{debug, error};
+use tracing::{debug, error, info};
+
+// Helper type to denote the fact that a Inbox that has received no checkpoint is different from an Inbox that has been checkpointed with checkpoint index 0
+#[derive(Debug, Copy, Clone)]
+pub struct OnchainCheckpointIndex {
+    index: Option<u32>,
+}
+
+impl OnchainCheckpointIndex {
+    pub fn from_checkpoint(checkpoint: &Checkpoint) -> Self {
+        Self {
+            index: if checkpoint.root.is_zero() {
+                None
+            } else {
+                Some(checkpoint.index)
+            },
+        }
+    }
+
+    pub fn from_index(index: u32) -> Self {
+        Self { index: Some(index) }
+    }
+
+    pub fn matches_signed_checkpoint(&self, other_checkpoint_index: u32) -> bool {
+        match self.index {
+            Some(index) => index >= other_checkpoint_index,
+            None => false,
+        }
+    }
+
+    pub fn is_behind_prover(&self, prover_count: usize) -> bool {
+        match self.index {
+            Some(index) => prover_count > index as usize + 1,
+            None => prover_count > 0,
+        }
+    }
+
+    pub fn next_leaf_index(&self) -> u32 {
```

Comment:
any feeling on specifying leaf here or not? I'm just noticing the rest of the impl doesn't mention leaves, so `next_index` may be more intuitive